### ISSUE TITLE
How to create a VPN User on GW with ELB

### ISSUE THREAD

**Issue by jalvarezferr** [AviatrixSystems/terraform-provider-aviatrix#98]
After updating the Terraform provider, the VPN User definition does not allow to provide the GW id when the GW is ELB enabled. Error message states that you can attach users to the LB but not the GW itself. However, the Gateway resource does not expose the elb_name property to Terraform, so there is no way to create a VPN user now.

Error is:

* aviatrix_vpn_user.coe02: failed to create Aviatrix VPNUser: This gateway
 is part of the Load Balancer Aviatrix-vpc-0ec7c0090x12c52b1c. You can add users
 to the Load Balancer but not to the gateway itself

If trying to get elb_name from the gw resource:

* aviatrix_vpn_user.coe02: Resource 'aviatrix_gateway.coe-eu-centra
l-1' does not have attribute 'elb_name' for variable 'aviatrix_gateway.coe-eu-central-1.elb_name'

**Comment by arvindsree** [comment]
[`Doc`](https://github.com/AviatrixSystems/terraform-provider-aviatrix/blob/master/website/docs/r/aviatrix_vpn_user.html.markdown)

The gw_name field should use ELB name(Aviatrix-vpc-0ec7c0090x12c52b1c) instead of the gateway name

**Comment by arvindsree** [comment]
This is a limitation of the REST API. You will have to create the gateway with a custom elb_name input

**Comment by lanmalkieri** [comment]
@arvindsree this still does not work (specifying a custom gateway name). When doing that we get this error. 

* aviatrix_vpn_user.this.10: failed to create Aviatrix VPNUser: This gateway is part of the Load Balancer calm-vpn. You can add users to the Load Balancer but not to the gateway itself

Which prevents creation still....

**Comment by arvindsree** [comment]
You need to use "calm-vpn" as  the gateway name in your case

**Comment by lanmalkieri** [comment]
That still does not work. It gives the exact same error stated above

```aviatrix_vpn_user.coe02: failed to create Aviatrix VPNUser: This gateway
is part of the Load Balancer Aviatrix-vpc-0ec7c0090x12c52b1c. You can add users
to the Load Balancer but not to the gateway itself```

Except obviously with the names filled in. When you create a custom name elb you still cannot specify either vpc id or elb name without it breaking.

**Comment by lanmalkieri** [comment]
The only way this works currently is if you don't create an ELB with the gateway. Please test on your side and validate. 

I am running tf version 0.1.11.

**Comment by jalvarezferr** [comment]
The limitation in fact is not of the API, which effectively return the elb_dns_name attribute of the gateway (from which the first 31 characters are the elb_name value), but of this provider.
I'm working on a modification of the provider to make it return the elb_name and elb_dns_name.

**Comment by jalvarezferr** [comment]
> The only way this works currently is if you don't create an ELB with the gateway. Please test on your side and validate.
> 
> I am running tf version 0.1.11.

It is a weird TF version, isn't it? My version (from terraform -v command) is v0.11.7

**Comment by jalvarezferr** [comment]
> The only way this works currently is if you don't create an ELB with the gateway. Please test on your side and validate.
> 
> I am running tf version 0.1.11.

You probably meant v0.11.11 (latest for now)

**Comment by lanmalkieri** [comment]
Yes I meant 0.11.11, thanks.

**Comment by arvindsree** [comment]
If you have not created the gateway 

resource "aviatrix_gateway" "vpn-gateway-test" {
 cloud_type = 1
 account_name = "aws" 
 gw_name = "gw1"
 vpc_id = "vpc-xxxxxxx"
 vpc_reg = "us-east-2"
 vpc_size = "t2.micro"
 vpn_access = "yes"
 cidr = "192.168.43.0/24"
 enable_elb = "yes"
 elb_name = "elb1"
 split_tunnel = "yes"
 vpc_net = "172.31.0.0/20"

 provisioner "local-exec" {
  command = "sleep 10"
 }
}

resource "aviatrix_vpn_user" "user1" {
 vpc_id = "vpc-xxxxxx"
 gw_name = "${aviatrix_gateway.vpn-gateway-test.elb_name}"
 user_name = "user1"
}


If you have already created the gateway earlier or outside of terraform or earlier (Not that update of elb_name is not suported)

If you have a custom LB name

resource "aviatrix_vpn_user" "user2" {
 vpc_id = "vpc-xxxxxx"
 gw_name = "elb1"
 user_name = "user2"
}

If the controller auto generated LB name,

resource "aviatrix_vpn_user" "user3" {
 vpc_id = "vpc-xxxxxx"
 gw_name = "Aviatrix-vpc-vpc-xxxxxx"
 user_name = "user3"
}

**Comment by jalvarezferr** [comment]
The fix is incomplete. Now that elb_name is computed, it requires a schema migration function that takes existing resources in the state and adds the elb_name to those that do not have it present, so it can be retrieved and used in vpn_user resources.
Finally, this fix should be backported to the 3.5 branch as it is a bug there. As the API in that version does not return the "lb_name" field it does now, it could be obtained from Agteway.ElbDNSName taking the initial part of the DNS name assigned by AWS.
